subsys/emds/emds.c

```diff
@@ -0,0 +1,327 @@
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		*size += NRFX_CEIL_DIV(ch->len, block_size) * block_size;
+		*size += NRFX_CEIL_DIV(emds_flash.ate_size, block_size) * block_size;
+		entries++;
+	}
+
+	struct emds_dynamic_entry *ch;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		*size += NRFX_CEIL_DIV(ch->entry.len, block_size) * block_size;
+		*size += NRFX_CEIL_DIV(emds_flash.ate_size, block_size) * block_size;
+		entries++;
+	}
+
+	return entries;
+}
+
+int emds_init(emds_store_cb_t cb)
+{
+	int rc;
+
+	if (emds_initialized) {
+		LOG_DBG("Already initialized");
+		return -EALREADY;
+	}
+
+	rc = emds_fs_init();
+	if (rc) {
+		return rc;
+	}
+
+	sys_slist_init(&emds_dynamic_entries);
+
+	app_store_cb = cb;
+
+	emds_initialized = true;
+
+	LOG_DBG("Emergency Data Storage initialized.");
+
+	return 0;
+}
+
+int emds_entry_add(const struct emds_entry *entry)
+{
+	struct emds_dynamic_entry *ch;
+	struct emds_dynamic_entry *item;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		if (ch->entry.id == entry->id) {
+			return -EINVAL;
+		}
+	}
+
+	item = k_malloc(sizeof(struct emds_dynamic_entry));
+	if (!item) {
+		return -ENOSPC;
+	}
+
+	item->entry = *entry;
+	sys_slist_append(&emds_dynamic_entries, &item->node);
+
+	emds_ready = false;
+
+	return 0;
+}
+
+int emds_store(void)
+{
+	if (!emds_ready) {
+		return -ECANCELED;
+	}
+
+	/* Lock all interrupts */
+	store_key = irq_lock();
+
+	/* Start the emergency data storage process. */
+	k_sem_give(&emds_sem);
+
+	return 0;
+}
+
+int emds_load(void)
+{
+	struct emds_dynamic_entry *ch;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		ssize_t len = emds_flash_read(&emds_flash,
+					      ch->entry.id, ch->entry.data,
+					      ch->entry.len);
+
+		if (len < 0) {
+			if (len != -ENXIO) {
+				LOG_ERR("Read dynamic entry: (%d) error (%d)",
+					ch->entry.id, len);
+			}
+		} else if (len != ch->entry.len) {
+			LOG_WRN("Read dynamic entry: (%d) did not match (%d:%d).",
+				ch->entry.id, ch->entry.len, len);
+		}
+	}
+
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		ssize_t len = emds_flash_read(&emds_flash,
+					      ch->id, ch->data, ch->len);
+
+		if (len < 0) {
+			if (len != -ENXIO) {
+				LOG_ERR("Read static entry: (%d) error (%d)",
+					ch->id, len);
+			}
+		} else if (len != ch->len) {
+			LOG_WRN("Read static entry: (%d) entry did not match (%d:%d)",
+				ch->id, ch->len, len);
+		}
+	}
+
+	return 0;
+}
+
+int emds_clear(void)
+{
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	return emds_flash_clear(&emds_flash);
+}
+
+int emds_prepare(void)
+{
+	uint32_t size;
+	int rc;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	(void)emds_entries_size(&size);
+
+	rc = emds_flash_prepare(&emds_flash, size);
+	if (rc) {
+		return rc;
+	}
+
+	emds_ready = true;
+
+	return 0;
+}
+
+uint32_t emds_store_time_get(void)
```

Comment:
This means that we have to add all the values that are in the datasheet. It also requires that we keep them updated when new devices comes. I would rather let the user actually go to the datasheet that they are using and then add this values them self based on what they are using. This will just give an estimation.

Reply:
Can you write this in the Kconfig entry description? Also, specify where the default value comes from (i.e. which device's datasheet).

Reply:
Added where this can be found in the documentation.